// wisp/engine/browser.js
{
    var _ns_ = {
            id: 'runner.main',
            doc: void 0
        };
    var wisp_compiler = require('wisp/compiler');
    var compile = wisp_compiler.compile;
}
var _wisp_runtime = exports._wisp_runtime = require('../runtime.js');
var _wisp_sequence = exports._wisp_sequence = require('../sequence.js');
var _wisp_string = exports._wisp_string = require('../string.js');
var fetchSource = exports.fetchSource = function fetchSource(src, callback) {
        return function () {
            var xhrø1 = new XMLHttpRequest();
            xhrø1.open('GET', src, true);
            xhrø1.addEventListener('load', function (ev) {
                return xhrø1.status >= 200 && xhrø1.status < 300 ? callback(xhrø1.responseText) : console.error(xhrø1.statusText);
            }, false);
            xhrø1.overrideMimeType ? xhrø1.overrideMimeType('text/plain') : void 0;
            xhrø1.setRequestHeader('If-Modified-Since', 'Fri, 01 Jan 1960 00:00:00 GMT');
            return xhrø1.send(null);
        }.call(this);
    };
var runWispCode = exports.runWispCode = function runWispCode(code, url) {
        return function () {
            var resultø1 = compile(code, { 'source-uri': url || 'inline' });
            var errorø1 = (resultø1 || 0)['error'];
            return errorø1 ? console.error(errorø1) : Function(eval((resultø1 || 0)['code']))();
        }.call(this);
    };
var fetchAndRunWispCode = exports.fetchAndRunWispCode = function fetchAndRunWispCode(url) {
        return fetchSource(url, function (code) {
            return runWispCode(code, url);
        });
    };
var __main__ = exports.__main__ = function __main__(ev) {
        [
            _wisp_string,
            _wisp_sequence,
            _wisp_runtime
        ].forEach(function (f) {
            return Object.keys(f).forEach(function ($1) {
                return (window || 0)[$1] = (f || 0)[$1];
            });
        });
        return function () {
            var scriptsø1 = document.getElementsByTagName('script');
            return function loop() {
                var recur = loop;
                var xø1 = 0;
                do {
                    recur = xø1 < scriptsø1.length ? function () {
                        var scriptø1 = (scriptsø1 || 0)[xø1];
                        var sourceø1 = scriptø1.src;
                        var contentø1 = scriptø1.text;
                        var contentTypeø1 = scriptø1.type;
                        contentTypeø1 == 'application/wisp' ? (function () {
                            sourceø1 ? fetchAndRunWispCode(sourceø1) : void 0;
                            return contentø1 ? runWispCode(contentø1, sourceø1) : void 0;
                        })() : void 0;
                        return loop[0] = xø1 + 1, loop;
                    }.call(this) : void 0;
                } while (xø1 = loop[0], recur === loop);
                return recur;
            }.call(this);
        }.call(this);
    };
window.addEventListener('load', __main__, false);
